### ISSUE TITLE
NIO version of CNV WDLs.

### ISSUE THREAD

**Issue by samuelklee** [broadinstitute/gatk#4806]

**Comment by LeeTL1220** [comment]
@samuelklee @bshifaw Please correct me if I missed a conversation etc....

Are you sure you want to do this before we have a portable WDL-NIO solution?  Adding NIO for the CNV workflow provides wins:
- for the reference fasta.
- in cases where we are only using a small portion of the bam files.

But for the latter point, this is not a common use case, right?
As far as I knew, the CNV WDLs were never using NIO, which seemed fine, for now.

**Comment by LeeTL1220** [comment]
Something is discussed in the style guide about NIO.

**Comment by samuelklee** [comment]
What is the timeline for portable WDL-NIO?

This would make things like performing preliminary analyses on subsets of contigs, etc. go a bit faster.  I agree that this is not a common use case, but since it's such a small amount of work, I don't see the harm.  Would be nice to be consistent with other Featured WDLs, if they're all using NIO as well (is this true?)

**Comment by samuelklee** [comment]
For example, I'm running some WGS tests on chr20-22.  The BAMs take ~1hr+ to localize but it only takes ~5 minutes to collect read/allelic counts.  Getting preempted adds a huge amount of unnecessary overhead.

**Comment by bshifaw** [comment]
Sure, but i'm not familiar with what the supported files would be. Would this be any input, reference, and resources files in google bukcets? 
Also, just glancing over the mutect2 nio looks like `File` is changed to `String` only in the task block, true?

**Comment by samuelklee** [comment]
Some of the files in the CNV workflows are not yet supported by NIO, so it won’t be as easy as changing File to String everywhere.  I think the ref and the BAMs are the most important ones to change, as Lee pointed out.

Let’s not worry too much about getting this in the next release or anything like that—-as soon as is convenient is fine.  But let’s decide on things like whether we need a copy in this repo and perhaps have a general strategy going forward.

**Comment by droazen** [comment]
@samuelklee Let us know if you want a quick consult on converting your tools to NIO -- happy to help!

**Comment by samuelklee** [comment]
Yup, this is tied in to the discussion of tabular data we touched on in https://github.com/broadinstitute/gatk/issues/4717.

CNV tabular data/files are represented as metadata (which is a SAM header that includes sample name and/or sequence dictionary) and records (which may or may not be locatable).  We can probably make do with just NIO support for locatable data, since these tend to be the largest and are currently the ones that are needlessly localized.

Let's discuss more in person!

**Comment by samuelklee** [comment]
@LeeTL1220 Actually, since the coverage collection itself is comparable to the time it takes to localize the BAM, this probably also helps even when you're not just running on snippets.  

Building a PoN on chr20-22 took ~2.5 hours without NIO (3/50 BAMs were preempted, some more than once), but it only took ~40 minutes with an NIO WDL I quickly whipped up (and 1/4 of that was because I was unlucky enough to get preempted on one BAM after 10 minutes)!

**Comment by sooheelee** [comment]
FYI, next release of Cromwell is supposed to automatically interpret File types as Sting when given a gs:// bucket file.

**Comment by LeeTL1220** [comment]
That would be really bad.  Are you sure?

**Comment by samuelklee** [comment]
Actually, I'm noticing that while using NIO for the BAM for read/allelic-count collection is usually much more efficient, using NIO for the reference in other tasks can be much slower.  Perhaps the access patterns for the reference (hitting ~10^5-10^6 intervals or sites for WES PreprocessIntervals and WES/WGS CollectAllelicCounts, respectively) make localization a better strategy?  @droazen does that sound right to you?

**Comment by LeeTL1220** [comment]
@sooheelee  I just confirmed that is *not* the case.  Have not confirmed the exact change yet.

**Comment by sooheelee** [comment]
There is a video recording of Chris L saying so. I’ll get the link to it in a bit.

**Comment by cjllanwarne** [comment]
There's a change in progress that will let you "opt in" to have Cromwell treat a specific `File` input as a `String` for the purposes of localization and as a `File` for the purposes of call caching. It'll probably look something like (but this isn't necessarily the final syntax final yet) this:
```wdl
task foo {
  input {
    File nio_me
    File dont_nio_me
  }
  parameter_meta {
    nio_me: {
      leave_in_cloud: true
    }
  }

  command {
    cat ~{dont_nio_me}
    java -jar gatk.jar --with-nio -i ~{nio_me}
  }
}
```

**Comment by cjllanwarne** [comment]
But this is still in progress so won't be in Cromwell 32 releasing today

**Comment by sooheelee** [comment]
Thanks @cjllanwarne for the clarification. @LeeTL1220, are you referring to the same feature or perhaps something else?

**Comment by LeeTL1220** [comment]
I have talked with Chris about this before and his post captures what I
understood.  This is not what I gathered from your previous post. Happy
that plans have not changed.

**Comment by sooheelee** [comment]
@LeeTL1220, Chris is also my source of information. Seems there's been a misunderstanding of what is meant by "automatic" in that we will have to opt-in to the feature.

**Comment by LeeTL1220** [comment]
@sooheelee Yeah, but we (WDL authors) need to opt-in per parameter.  There are still files that we would want to localize even if a gs url is provided.

**Comment by sooheelee** [comment]
@cjllanwarne, how do I opt-in per parameter? I am trying to run the v4.0.11.0 gCNV WDL using Cromwell v36 and even after changing `File` type to `String` type for the relevant BAM/BAI inputs, I am getting a file-type not recognized error.

**Comment by sooheelee** [comment]
I spoke too soon. No matter how I define the type, whether String or File, I get the error:

```
ERROR: Value for this attribute is expected to be a string:

        bam: {
```

I added the following parameter_meta field to the task:

![screenshot 2018-11-08 18 02 03](https://user-images.githubusercontent.com/11543866/48232693-7569ff00-e380-11e8-9dad-2eed3ca68118.png)

How to correct this @cjllanwarne?

**Comment by samuelklee** [comment]
@sooheelee I found that on FC, it was only necessary to make the File -> String change at the task level (i.e., in CollectCounts).  Not sure if this works due to the particular version of Cromwell on FC.  In any case, I don't think you should stress too much about getting NIO working on your VM.  Again, I'd expect the current non-NIO gCNV WDL to only take a few hours on FC, and then less than that once coverage collection is call cached.  No use spending more time getting NIO working than it would save you, after all!

**Comment by sooheelee** [comment]
I'm informed that I have access to the broad-firecloud-dsde billing account and it appears whatever error we encountered the other day was something transient.

**Comment by mwalker174** [comment]
NIO streaming is currently used to avoid BAM localization in `CollectCounts` and `CollectAllelicCounts`, which is particularly beneficial when running on a limited set of intervals. It is also used in `JointSegmentation` to stream VCFs, which saves on disk space.

Additional tasks that could benefit from streaming are those that use counts files inputs: `FilterIntervals`, `GermlineCNVCallerCaseMode`, `GermlineCNVCallerCohortMode`, `DetermineGermlineContigPloidyCaseMode`, and `DetermineGermlineContigPloidyCohortMode`, `CreateReadCountPanelOfNormals`, and `DenoiseReadCounts`. These would require that read counts are in TSV format, which we should move to using exclusively (instead of HDF5).

**Comment by samuelklee** [comment]
@asmirnov239 note https://github.com/broadinstitute/gatk/pull/6266.
